src/bench/pqs/spraylist.cpp

```diff
@@ -55,14 +55,38 @@ void
 spraylist::insert(const uint32_t &k,
                   const uint32_t &v)
 {
-    sl_add(m_q, k, READ_ADD_REM_ELASTIC_TX);
+    sl_add_val(m_q, k, v, TRANSACTIONAL);
+}
+
+void
+spraylist::insert(const size_t &k,
+                  const size_t &v)
+{
+    sl_add_val(m_q, k, v, TRANSACTIONAL);
 }
 
 bool
-spraylist::delete_min(uint32_t &v)
+spraylist::delete_min(size_t &k, size_t &v)
 {
-    const int ret = spray_delete_min(m_q, &v, d);
-    return (ret == 1);
+    unsigned long k_ret;
+    unsigned long v_ret;
+    int ret;
+    do {
+        ret = spray_delete_min_key(m_q, &k_ret, &v_ret, d);
```

Comment:
You're right, there is a retry loop here: https://github.com/klsmpq/klsm/blob/master/src/k_lsm/k_lsm_inl.h#L104

It retries if both the local dlsm and the shared slsm are empty, and spying (taking items from another thread's local dlsm) copied more than 0 items to the local dlsm.

It differs from Listing 5 in that failed CAS result in a spurious failure.

I'm fine with keeping the SprayList loop if you think that is appropriate (it'd be great to have a short comment as to what the conditions (return value, key value) mean).